### PR TITLE
Add SCIM2 Custom schema support

### PR DESCRIPTION
**Pull request by piraveena**
Resolves https://github.com/wso2/product-is/issues/11705

Need to merge charon PR and framework PR before this annd bump their versions here

### CODE REVIEW

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/handlers/SCIMClaimOperationEventHandler.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    private static final Log log = LogFactory.getLog(SCIMClaimOperationEventHandler.class);
+    public static final String WSO2_CARBON_DIALECT = "http://wso2.org/claims";
+
+    @Override
+    public void handleEvent(Event event) throws IdentityEventException {
+
+        int tenantId = (int) event.getEventProperties().get(IdentityEventConstants.EventProperty.TENANT_ID);
+        if (log.isDebugEnabled()) {
+            log.debug(event.getEventName() + " event received to SCIMClaimOperationEventHandler for the tenant with " +
+                    "Id: "+ tenantId);
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/handlers/SCIMClaimOperationEventHandler.java:44]: formatting issue

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/handlers/SCIMClaimOperationEventHandler.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+import org.wso2.carbon.identity.event.IdentityEventConstants;
+import org.wso2.carbon.identity.event.IdentityEventException;
+import org.wso2.carbon.identity.event.event.Event;
+import org.wso2.carbon.identity.event.handler.AbstractEventHandler;
+import org.wso2.carbon.identity.scim2.common.cache.SCIMCustomAttributeSchemaCache;
+import org.wso2.carbon.identity.scim2.common.utils.SCIMCommonUtils;
+
+import static org.wso2.carbon.identity.scim2.common.utils.SCIMCommonUtils.getCustomSchemaURI;
+
+public class SCIMClaimOperationEventHandler extends AbstractEventHandler {
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/handlers/SCIMClaimOperationEventHandler.java:33]: let's add a class level comment

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/IdentityResourceTypeResourceManager.java [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+                String error = "Newly created User resource is null.";
+                throw new InternalErrorException(error);
+            }
+            //put the uri of the resource type object in the response header parameter.
+            return new SCIMResponse(ResponseCodeConstants.CODE_OK,
+                    encodedObject, responseHeaders);
+        } catch (CharonException | BadRequestException | InternalErrorException | NotFoundException e) {
+            return encodeSCIMException(e);
+        } catch (JSONException e) {
+            return null;
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/IdentityResourceTypeResourceManager.java:127]: shall we log the exception before returning

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 4]
```diff
@@ -344,6 +346,8 @@ public User createUser(User user, Map<String, Boolean> requiredAttributes)
                 throw new BadRequestException(errorMessage, ResponseCodeConstants.INVALID_VALUE);
             }
             handleErrorsOnUserNameAndPasswordPolicy(e);
+        } catch (BadRequestException | NotImplementedException e) {
+            throw new CharonException("Error in getting user information from Carbon User Store", e);
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:350]: let;s log the exception before rethrowing

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 5]
```diff
@@ -2010,6 +2022,11 @@ private Condition getCondition(Node node, Map<String, String> attributes) throws
                         attributes.put(claim.getClaim().getClaimUri(), claim.getMappedAttribute(domainName));
                     }
                 }
+                if (customClaims != null) {
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:2025]: let's use ArrayUtils

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java [hunk 6]
```diff
@@ -5552,4 +5579,50 @@ private String getMultivaluedAttributeSeparator(String userStoreDomainName) {
...
+     * Return list of attributes in the custom schema of the tenant.
+     *
+     * @return Return list of attributes.
+     * @throws CharonException
+     */
+    @Override
+    public List<Attribute> getCustomUserSchemaAttributes() throws CharonException {
+
+        if (!SCIMCommonUtils.isCustomSchemaEnabled()) {
+            log.error("Custom schema is disabled in server level");
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/SCIMUserManager.java:5593]: shall we make this a debug log

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/AttributeMapper.java [hunk 7]
```diff
@@ -245,16 +247,9 @@ private static void setClaimsForComplexAttribute(Attribute entry, Map<String, St
...
-     * Construct the SCIM Object given the attribute URIs and attribute values of the object.
-     *
-     * @param attributes
-     * @param scimObjectType
-     * @return
-     */
-    public static SCIMObject constructSCIMObjectFromAttributes(Map<String, String> attributes,
-                                                               int scimObjectType)
-            throws CharonException, NotFoundException, BadRequestException {
+    public static SCIMObject constructSCIMObjectFromAttributes(UserManager userManager, Map<String, String> attributes,
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/AttributeMapper.java:250]: don't change public methods. use overloaded methods instead

File: components/org.wso2.carbon.identity.scim2.provider/src/main/java/org/wso2/carbon/identity/scim2/provider/util/SupportUtils.java [hunk 8]
```diff
@@ -79,4 +101,62 @@ public static String getAuthenticatedUsername() {
...
+    /**
+     * This builds the custom schema for tenants.
+     * @param userManager
+     * @param tenantId
+     * @throws CharonException
+     */
+    public static void buildCustomSchema(UserManager userManager, int tenantId) throws CharonException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Building scim2 custom attribute schema for tenant with Id: "+ tenantId);
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.scim2.provider/src/main/java/org/wso2/carbon/identity/scim2/provider/util/SupportUtils.java:114]: formatting issue

File: components/org.wso2.carbon.identity.scim2.provider/src/main/java/org/wso2/carbon/identity/scim2/provider/util/SupportUtils.java [hunk 9]
```diff
@@ -79,4 +101,62 @@ public static String getAuthenticatedUsername() {
...
+        } catch (NotImplementedException | BadRequestException | IdentitySCIMException e) {
+            throw new CharonException("Error while building scim custom schema", e);
+        }
+    }
+
+    public static int getTenantId() {
+
+        String tenantDomain = getTenantDomain();
+        if (StringUtils.isNotBlank(tenantDomain)) {
+            return IdentityTenantUtil.getTenantId(tenantDomain);
```
Comment (emswbandara) [inline components/org.wso2.carbon.identity.scim2.provider/src/main/java/org/wso2/carbon/identity/scim2/provider/util/SupportUtils.java:149]: AFAIK IdentityTenantUtil.getTenantId already handles the tenantdomain being null case. Please check

File: pom.xml [hunk 10]
```diff
@@ -301,11 +301,11 @@
         <inbound.auth.oauth.version>6.5.3</inbound.auth.oauth.version>
         <commons-collections.version>3.2.0.wso2v1</commons-collections.version>
         <carbon.kernel.version>4.6.2-m9</carbon.kernel.version>
-        <identity.framework.version>5.20.33</identity.framework.version>
+        <identity.framework.version>5.20.51-SNAPSHOT</identity.framework.version>
```
Comment (emswbandara) [inline pom.xml:304]: remove snapshot versions when other component PRs get merged

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/handlers/SCIMClaimOperationEventHandler.java (hunk 2 above)
Comment (piraveena) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/handlers/SCIMClaimOperationEventHandler.java:36]: fixed with https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/352/commits/9f6d7fe7474928c000edfe7c28b6e6b84cb777c7

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/IdentityResourceTypeResourceManager.java (hunk 3 above)
Comment (piraveena) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/impl/IdentityResourceTypeResourceManager.java:136]: fixed with https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/352/commits/9f6d7fe7474928c000edfe7c28b6e6b84cb777c7

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/AttributeMapper.java (hunk 7 above)
Comment (piraveena) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/AttributeMapper.java:250]: It is a overloaded method. You can find the previous method here. https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/blob/e9cd1c6e31066bcb2d3764a59d851688c05f746d/components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/AttributeMapper.java#L384.

File: pom.xml (hunk 10 above)
Comment (piraveena) [reply pom.xml:304]: yes. need to bump charon and framework versions

**Comment by JKAUSHALYA** [comment]
What is the reason for build failure?

**Comment by piraveena** [comment]
> What is the reason for build failure?

@JKAUSHALYA this depends on Charon PR and framework PR. Once they got merged, I need to bump the respective versions

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/handlers/SCIMClaimOperationEventHandler.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+            if (log.isDebugEnabled()) {
+                log.debug("SCIM2 Custom user schema has disabled in server level.");
+            }
+            return;
+        }
+
+        String claimDialectUri =
+                (String) event.getEventProperties().get(IdentityEventConstants.EventProperty.CLAIM_DIALECT_URI);
+        if (StringUtils.isNotBlank(claimDialectUri) && !(claimDialectUri.equalsIgnoreCase(getCustomSchemaURI())
+                || claimDialectUri.equalsIgnoreCase(WSO2_CARBON_DIALECT))) {
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/handlers/SCIMClaimOperationEventHandler.java:60]: is this claimDialectUri.equalsIgnoreCase(WSO2_CARBON_DIALECT) or !claimDialectUri.equalsIgnoreCase(WSO2_CARBON_DIALECT) ?

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/handlers/SCIMClaimOperationEventHandler.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+            if (log.isDebugEnabled()) {
+                log.debug("Needs to handle only if this claim update happens to SCIM2 custom schema or local claim "
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/handlers/SCIMClaimOperationEventHandler.java:62]: Change the debug message in a more meaningful manner.

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/handlers/SCIMClaimOperationEventHandler.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+/**
+ * This handles the claim metadata operation related events.
+ */
+public class SCIMClaimOperationEventHandler extends AbstractEventHandler {
+
+    private static final Log log = LogFactory.getLog(SCIMClaimOperationEventHandler.class);
+    public static final String WSO2_CARBON_DIALECT = "http://wso2.org/claims";
+
+    @Override
+    public void handleEvent(Event event) throws IdentityEventException {
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/handlers/SCIMClaimOperationEventHandler.java:42]: Explain the job of this handler as a comment.

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/handlers/SCIMClaimOperationEventHandler.java (hunk 11 above)
Comment (piraveena) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/handlers/SCIMClaimOperationEventHandler.java:60]: I'm using as 

if (StringUtils.isNotBlank(claimDialectUri) && ! **(claimDialectUri.equalsIgnoreCase(getCustomSchemaURI())
                || claimDialectUri.equalsIgnoreCase(WSO2_CARBON_DIALECT))**) , then return.

if it is not a customDialect claim or localclaimdialect, then we don't need to handle it

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/handlers/SCIMClaimOperationEventHandler.java (hunk 13 above)
Comment (piraveena) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/handlers/SCIMClaimOperationEventHandler.java:52]: added to class level

**Comment by piraveena** [comment]
These are the two test cases got failed. 

```
ERROR] Tests run: 1671, Failures: 2, Errors: 0, Skipped: 5, Time elapsed: 5,462.835 s <<< FAILURE! - in TestSuite
[ERROR] testCreateUser(org.wso2.identity.integration.test.scim2.rest.api.SCIMUserUpdateTest)  Time elapsed: 0.196 s  <<< FAILURE!
java.lang.AssertionError: expected:<2> but was:<3>
	at org.wso2.identity.integration.test.scim2.rest.api.SCIMUserUpdateTest.testCreateUser(SCIMUserUpdateTest.java:134)

[ERROR] testCreateUser(org.wso2.identity.integration.test.scim2.rest.api.SCIMUserUpdateTest)  Time elapsed: 0.171 s  <<< FAILURE!
java.lang.AssertionError: expected:<2> but was:<3>
	at org.wso2.identity.integration.test.scim2.rest.api.SCIMUserUpdateTest.testCreateUser(SCIMUserUpdateTest.java:134)

[INFO
```
This is due to the nnumber of schemas has increased in the scim response. It is fixed with this PR https://github.com/wso2/product-is/pull/11769

**Review by IsuraD — APPROVED** [review]

**Pull request merged**